src/main/java/is/citizen/sdk/crypto/CitizenCrypto.java

```diff
@@ -0,0 +1,364 @@
+import java.security.spec.InvalidKeySpecException;
+import java.security.spec.X509EncodedKeySpec;
+import java.util.*;
+
+public class CitizenCrypto {
+    private static final String CRYPTO_ALGORITHM = "RSA";
+    private static final int CRYPTO_KEY_SIZE = 2048;
+    private static final String CRYPTO_PRIVATE_KEY_ENCRYPTION_ALGO = "AES-256-GCM";
+    private static final String AUTH_ALGORITHM = "ECDSA";
+    private static final String AUTH_CURVE = "secp256r1";
+    private static final String AUTH_SIGNATURE_ALGORITHM = "SHA256withECDSA";
+    private static final String BOUNTY_CASTLE_PROVIDER = "BC";
+
+    static {
+        if (Security.getProvider(BOUNTY_CASTLE_PROVIDER) == null) {
+            Security.addProvider(new BouncyCastleProvider());
+        }
+    }
+
+    public static Optional<KeyHolder> generateCryptoKeyPair(String secret) {
+        KeyHolder keyHolder = null;
+        try {
+            KeyPairGenerator keyGen = KeyPairGenerator.getInstance(CRYPTO_ALGORITHM, BOUNTY_CASTLE_PROVIDER);
+            keyGen.initialize(CRYPTO_KEY_SIZE);
+            KeyPair keyPair = keyGen.generateKeyPair();
+
+            String privateKey = encryptPrivateKey(keyPair.getPrivate(), secret);
+
+            String publicKey = getKeyAsString(keyPair.getPublic());
+
+            keyHolder = new KeyHolder();
+
+            keyHolder.setPrivateKey(privateKey);
+            keyHolder.setPublicKey(publicKey);
+        } catch (IOException | NoSuchAlgorithmException | NoSuchProviderException e) {
+            throw new CryptoException("CitizenCrypto exception: " + e.getMessage());
+        }
+
+        return Optional.ofNullable(keyHolder);
+    }
+
+    public static Optional<KeyHolder> generateAuthenticationKeyPair(String secret) {
+        KeyHolder keyHolder = null;
+        try {
+            ECNamedCurveParameterSpec ecNamedCurveParameterSpec = ECNamedCurveTable.getParameterSpec(AUTH_CURVE);
+            KeyPairGenerator keyGen = KeyPairGenerator.getInstance(AUTH_ALGORITHM, BOUNTY_CASTLE_PROVIDER);
+            keyGen.initialize(ecNamedCurveParameterSpec);
+            KeyPair keyPair = keyGen.generateKeyPair();
+
+            String privateKey = encryptPrivateKey(keyPair.getPrivate(), secret);
+
+            String publicKey = getKeyAsString(keyPair.getPublic());
+
+            keyHolder = new KeyHolder();
+
+            keyHolder.setPrivateKey(privateKey);
+            keyHolder.setPublicKey(publicKey);
+
+        } catch (IOException | NoSuchAlgorithmException | NoSuchProviderException | InvalidAlgorithmParameterException e) {
+            throw new CryptoException("CitizenCrypto exception: " + e.getMessage());
+        }
+
+        return Optional.ofNullable(keyHolder);
+    }
+
+    public static Optional<KeyHolder> generateSigningKeyPair(String secret) {
+        return generateAuthenticationKeyPair(secret);
+    }
+
+    public static String signData(byte[] dataToSign, String encodedPrivateKey, String password) {
```

Comment:
password -> secret. Let's keep the wording consistent.

Reply:
Fixed.